### PR TITLE
Add software models

### LINKED ISSUE #1: Add support for an OS Model to define target OS image per device
### Proposed Functionality
Support a model to define OSs that should be used by each device and ensure there is also a relationship between vendor and platform.

### Use Case
As a user, I should be able to define the intended OS or list of supported OSs per device. I should preferably be able to run search queries within the UI using greater than and less than (<>) operators to verify if my devices.

### Database Changes
Yes, will require a new database model(s), as well as updates to Vendor and Platform models

### External Dependencies
Maybe - need to explore external library(ies) for parsing idiosyncratic vendor OS version strings.

### PR DESCRIPTION
**Pull request by gsnider2195**
# Closes #1
# What's Changed

- Create SoftwareVersion model
- Create SoftwareImage model
- Add FK from Device to SoftwareVersion
- Add FK from VirtualMachine to SoftwareVersion
- Add FK from InventoryItem to SoftwareVersion
- Add m2m from DeviceType to SoftwareImage, with a boolean for "default image"

# Screenshots

## Nav menu

image

## SoftwareVersion list

image

## SoftwareVersion detail

image

## SoftwareImage list
image

## SoftwareImage detail
image

## DeviceType create/edit
image
## DeviceType detail
image

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Create SoftwareVersion model
  - [x] Migrations
  - [x] Default statuses (data migration)
  - [x] Serializer
  - [x] List view / Table
  - [x] Create/edit form
  - [x] Create/edit view
  - [x] Bulk edit form
  - [x] Bulk edit view
  - [x] Filterset
  - [x] Filter form
  - [x] Factory
  - [x] Unit tests
  - [x] Documentation
  - [x] Update nautobot.apps
- [x] Create SoftwareImage model
  - [x] Migrations
  - [x] Default statuses (data migration)
  - [x] Serializer
  - [x] List view / Table
  - [x] Create/edit form
  - [x] Create/edit view
  - [x] Bulk edit form
  - [x] Bulk edit view
  - [x] Filterset
  - [x] Filter form
  - [x] Factory
  - [x] Unit tests
  - [x] Documentation
  - [x] Update nautobot.apps
- [x] Add FK from Device to SoftwareVersion
  - [x] Migration
  - [x] Model validation (device type must match softwareimage and device)
  - [x] Update querysets (select_related)
  - [x] Serializer
  - [x] Create/edit form
  - [x] Create/edit view
  - [x] Bulk edit form
  - [x] Bulk edit view
  - [x] Device detail view
  - [x] SoftwareVersion detail view
  - [x] Filterset
  - [x] Filter form
  - [x] Factory
  - [x] Unit tests
  - [x] Documentation
- [x] Add FK from VirtualMachine to SoftwareVersion
  - [x] Migration
  - [x] Update querysets (select_related)
  - [x] Serializer
  - [x] Create/edit form
  - [x] Create/edit view
  - [x] Bulk edit form
  - [x] Bulk edit view
  - [x] VirtualMachine detail view
  - [x] SoftwareVersion detail view
  - [x] Filterset
  - [x] Filter form
  - [x] Factory
  - [x] Unit tests
  - [x] Documentation
- [x] Add FK from InventoryItem to SoftwareVersion
  - [x] Migration
  - [x] Update querysets (select_related)
  - [x] Serializer
  - [x] Create/edit form
  - [x] Create/edit view
  - [x] Bulk edit form
  - [x] Bulk edit view
  - [x] InventoryItem detail view
  - [x] SoftwareVersion detail view
  - [x] Filterset
  - [x] Filter form
  - [x] Factory
  - [x] Unit tests
  - [x] Documentation
- [x] Add m2m from DeviceType to SoftwareImage, with a boolean for "default image"
  - [x] Migration
  - [x] Model validation (device type may only have one default softwareimage)
  - [x] Update querysets (prefetch_related)
  - [x] Serializer
  - [x] Expose through table in API
    - [x] Serializer
    - [x] Filterset
    - [x] Viewset
    - [x] Tests
  - [x] Create/edit form
  - [x] Create/edit view
  - [x] Bulk edit form
  - [x] Bulk edit view
  - [x] DeviceType detail view
  - [x] SoftwareImage detail view
  - [x] Filterset
  - [x] Filter form
  - [x] Factory
  - [x] Unit tests
  - [x] Documentation
  - n/a Update nautobot.apps
- [ ] Release overview documentation
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/dcim/models/devices.py [hunk 1]
```diff
@@ -979,3 +985,116 @@ def devices_sorted(self):
...
+    "statuses",
+    "webhooks",
+)
+class SoftwareImage(PrimaryModel):
+    """A software image for a Device, Virtual Machine or Inventory Item."""
+
+    software_version = models.ForeignKey(
+        to="SoftwareVersion", on_delete=models.CASCADE, related_name="software_images", verbose_name="Software Version"
+    )
+    image_file_name = models.CharField(blank=False, max_length=100, verbose_name="Image File Name")
```
Comment (gsnider2195) [inline nautobot/dcim/models/devices.py:1026]: TODO: Add file size field

File: nautobot/dcim/templates/dcim/softwareimage_retrieve.html [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                        </div>
+                    </div>
+                    {% plugin_full_width_page object %}
+                </div>
+            </div>
+            <div class="row">
+                <div class="col-md-6">
+                    <div class="panel panel-default">
+                        <div class="panel-heading">
+                            <strong>Stats</strong>
```
Comment (gsnider2195) [inline nautobot/dcim/templates/dcim/softwareimage_retrieve.html:97]: TODO: Move to SoftwareVersion, add VirtualMachine

File: nautobot/dcim/templates/dcim/softwareimage_retrieve.html [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                                        {% if object.image_file_checksum %}
+                                        <span class="hover_copy">
+                                            <span id="copy_image_file_checksum">{{ object.image_file_checksum }}</span>
+                                            <button type="button" class="btn btn-inline btn-default hover_copy_button" data-clipboard-target="#copy_image_file_checksum">
+                                                <span class="mdi mdi-content-copy"></span>
+                                            </button>
+                                        </span>
+                                        {% else %}
+                                        <span class="text-muted">&mdash;</span>
+                                        {% endif %}
```
Comment (glennmatthews) [inline nautobot/dcim/templates/dcim/softwareimage_retrieve.html:55]: Would be really good to add a templatetag that renders this repeated pattern for copyable fields.

File: nautobot/dcim/templates/dcim/softwareimage_retrieve.html [hunk 4]
```diff
@@ -0,0 +1,174 @@
+{% extends 'generic/object_retrieve.html' %}
+{% load helpers %}
+{% load plugins %}
+{% load render_table from django_tables2 %}
+
+
+{% block content %}
```
Comment (glennmatthews) [inline nautobot/dcim/templates/dcim/softwareimage_retrieve.html:7]: Is there a reason we need to override the entire `content` block instead of just `content_left_page`/`content_right_page`/`content_full_width_page`?

File: nautobot/dcim/templates/dcim/softwareversion_retrieve.html [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+{% endblock content_left_page %}
+
+{% block content_right_page %}
+<div class="panel panel-default">
+    <div class="panel-heading">
+        <strong>Stats</strong>
+    </div>
+    <div class="panel-body">
+        <div class="col-md-4 text-center">
+            <h2><a href="{% url 'dcim:softwareimage_list' %}?software_version={{ object.pk }}" class="btn {% if object.software_images.count %}btn-primary{% else %}btn-default{% endif %} btn-lg">{{ object.software_images.count }}</a></h2>
```
Comment (glennmatthews) [inline nautobot/dcim/templates/dcim/softwareversion_retrieve.html:67]: Are we incurring any unnecessary repeated calculations by evaluating `.count` twice here? Should we stash it in a variable that we can reuse as needed instead?

File: nautobot/dcim/tests/test_api.py [hunk 6]
```diff
@@ -762,6 +767,9 @@ class DeviceTypeTest(APIViewTestCases.APIViewTestCase):
 
     @classmethod
     def setUpTestData(cls):
+        cls.VERBOTEN_STRINGS = tuple(
+            [o for o in cls.VERBOTEN_STRINGS if o not in SoftwareImageHashingAlgorithmChoices.as_dict().keys()]
+        )
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_api.py:772]: What does this do here and below? Mind adding a comment?

File: nautobot/extras/management/__init__.py [hunk 7]
```diff
@@ -42,6 +44,8 @@
     "Decommissioning": ColorChoices.COLOR_AMBER,
     "Deprecated": ColorChoices.COLOR_RED,
     "Deprovisioning": ColorChoices.COLOR_AMBER,
+    "End-of-life": ColorChoices.COLOR_RED,
```
Comment (glennmatthews) [inline nautobot/extras/management/__init__.py:47]: Should we change this to `End-of-Life` throughout?

File: nautobot/virtualization/forms.py [hunk 8]
```diff
@@ -242,6 +243,7 @@ class Meta:
         widgets = {
             "primary_ip4": StaticSelect2(),
             "primary_ip6": StaticSelect2(),
+            "software_version": StaticSelect2(),
```
Comment (glennmatthews) [inline nautobot/virtualization/forms.py:246]: Why are these needed? Shouldn't we just be explicitly using `DynamicModelChoiceField` for these three fields instead?

File: nautobot/virtualization/templates/virtualization/virtualmachine.html [hunk 9]
```diff
@@ -111,6 +111,16 @@
...
                     </td>
                 </tr>
+                <tr>
+                    <td>Software Version</td>
+                    <td>
+                        {% if object.software_version %}
+                            {{ object.software_version|hyperlinked_object }}
+                        {% else %}
+                            <span class="text-muted">&mdash;</span>
+                        {% endif %}
```
Comment (glennmatthews) [inline nautobot/virtualization/templates/virtualization/virtualmachine.html:121]:
```suggestion
                        {{ object.software_version|hyperlinked_object }}
```

File: nautobot/dcim/models/devices.py [hunk 10]
```diff
@@ -979,3 +1034,122 @@ def devices_sorted(self):
...
+    "webhooks",
+)
+class SoftwareImage(PrimaryModel):
+    """A software image for a Device, Virtual Machine or Inventory Item."""
+
+    software_version = models.ForeignKey(
+        to="SoftwareVersion", on_delete=models.CASCADE, related_name="software_images", verbose_name="Software Version"
+    )
+    image_file_name = models.CharField(blank=False, max_length=255, verbose_name="Image File Name")
+    image_file_checksum = models.CharField(blank=True, max_length=256, verbose_name="Image File Checksum")
```
Comment (HanlinMiao) [inline nautobot/dcim/models/devices.py:1085]: what is the reason for `max_length` differ here?

File: nautobot/dcim/models/devices.py [hunk 11]
```diff
@@ -979,3 +1034,122 @@ def devices_sorted(self):
...
+    status = StatusField(blank=False, null=False)
+
+    objects = BaseManager.from_queryset(SoftwareImageQuerySet)()
+
+    class Meta:
+        ordering = ("software_version", "image_file_name")
+        unique_together = ("image_file_name", "software_version")
+
+    def __str__(self):
+        return self.image_file_name
```
Comment (HanlinMiao) [inline nautobot/dcim/models/devices.py:1109]: Should we include `software_version` here as well since `image_file_name` is not unique by itself?

File: nautobot/dcim/templates/dcim/device_edit.html [hunk 12]
```diff
@@ -71,6 +71,7 @@
             {% render_field form.secrets_group %}
             {% render_field form.device_redundancy_group %}
             {% render_field form.device_redundancy_group_priority %}
+            {% render_field form.software_version %}
```
Comment (HanlinMiao) [inline nautobot/dcim/templates/dcim/device_edit.html:74]: Per DSU discussion, should we make this field up closer to `device_type`?

File: nautobot/dcim/templates/dcim/softwareversion_retrieve.html (hunk 5 above)
Comment (HanlinMiao) [reply nautobot/dcim/templates/dcim/softwareversion_retrieve.html:67]: Maybe with a `with` tag?

File: nautobot/dcim/tests/test_api.py [hunk 13]
```diff
@@ -2613,3 +2638,134 @@ def setUpTestData(cls):
...
+        DeviceTypeToSoftwareImage.objects.create(
+            device_type=device_types[0],
+            software_image=software_images[1],
+            is_default=False,
+        )
+        DeviceTypeToSoftwareImage.objects.create(
+            device_type=device_types[0],
+            software_image=software_images[2],
+            is_default=True,
+        )
```
Comment (HanlinMiao) [inline nautobot/dcim/tests/test_api.py:2750]: Seems like we are allowing two `is_default=True` software images on the same device type here.

File: nautobot/dcim/tests/test_models.py [hunk 14]
```diff
@@ -1155,6 +1158,55 @@ def test_primary_ip_validation_logic(self):
...
+        )
+
+        device_type_to_software_image = DeviceTypeToSoftwareImage(
+            device_type=device_type,
+            software_image=software_images[1],
+            is_default=True,
+        )
+
+        with self.assertRaises(ValidationError):
+            device_type_to_software_image.validated_save()
```
Comment (HanlinMiao) [inline nautobot/dcim/tests/test_models.py:1202]: Can you point me to where the uniqueness constraint is set here?

File: nautobot/core/templates/panel_table.html [hunk 15]
```diff
@@ -7,7 +7,9 @@
         </div>
     {% endif %}
     {% if table.rows %}
-        {% render_table table table_inc_template|default:'inc/table.html' %}
+        <div class="table-responsive">
+            {% render_table table table_inc_template|default:'inc/table.html' %}
+        </div>
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/core/templates/panel_table.html:12]: In places where `panel_table.html` is used. There are still some places where we're doing `{% render_table table 'inc/table.html' %}` that are still broken with this change. We probably need to just standardize on one way to render tables and fix it across the board.

File: nautobot/dcim/factory.py [hunk 16]
```diff
@@ -248,6 +266,17 @@ class Meta:
     has_comments = NautobotBoolIterator()
     comments = factory.Maybe("has_comments", factory.Faker("paragraph"), "")
 
+    has_software_images = NautobotBoolIterator()
+
+    @factory.post_generation
+    def software_images(self, create, extracted, **kwargs):
+        if not create or not DeviceTypeFactory.has_software_images.evaluate(None, None, None):
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/dcim/factory.py:273]: That's how factoryboy grabs the next value from a factory `Declaration`. `None, None, None` is required because evaluate needs 3 positional args but they're not doing anything in this context so I just used None.

File: nautobot/dcim/models/devices.py [hunk 17]
```diff
@@ -92,6 +98,25 @@ def __str__(self):
         return self.name
 
 
+@extras_features("graphql")
+class DeviceTypeToSoftwareImage(BaseModel, ChangeLoggedModel):
+    device_type = models.ForeignKey("dcim.DeviceType", on_delete=models.CASCADE, related_name="software_image_mappings")
+    software_image = models.ForeignKey(
+        "dcim.SoftwareImage", on_delete=models.PROTECT, related_name="device_type_mappings"
+    )
+    is_default = models.BooleanField(default=False, help_text="Is the default image for this device type")
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/dcim/models/devices.py:107]: Coming in the next commit. Thanks!

File: nautobot/dcim/models/devices.py [hunk 18]
```diff
@@ -979,3 +1034,122 @@ def devices_sorted(self):
...
+
+#
+# Software images
+#
+
+
+class SoftwareImageQuerySet(RestrictedQuerySet):
+    """Queryset for SoftwareImage objects."""
+
+    def get_for_object(self, obj):
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/dcim/models/devices.py:1056]: This is just a convenience method and not used for any views so it would probably be more overhead to do the invalidation?

File: nautobot/dcim/models/devices.py [hunk 19]
```diff
@@ -979,3 +1034,122 @@ def devices_sorted(self):
...
+        """Return all SoftwareVersion assigned to the given object."""
+        if not isinstance(obj, models.Model):
+            raise TypeError(f"{obj} is not an instance of Django Model class")
+        if isinstance(obj, Device):
+            qs = self.filter(platform__devices=obj)
+        elif isinstance(obj, InventoryItem):
+            qs = self.filter(platform__devices__inventory_items=obj)
+        else:
+            qs = self
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/dcim/models/devices.py:1124]: Same as above

File: nautobot/dcim/templates/dcim/softwareimage_retrieve.html (hunk 3 above)
Comment (gsnider2195) [reply nautobot/dcim/templates/dcim/softwareimage_retrieve.html:55]: I was going to write one but can't remember why I didn't..

File: nautobot/dcim/templates/dcim/softwareimage_retrieve.html (hunk 4 above)
Comment (gsnider2195) [reply nautobot/dcim/templates/dcim/softwareimage_retrieve.html:7]: `content_full_width_page` appears below `left_page` and `right_page` which means the main panel for this object would be below the `tags` panel and any other panels in `left/right_page`. This is completely copy/pasted just to move `content_full_width_page` above those other 2.

File: nautobot/core/management/commands/generate_test_data.py [hunk 20]
```diff
@@ -161,6 +167,8 @@ def _generate_factory_data(self, seed, db_name):
         DeviceRedundancyGroupFactory.create_batch(20, using=db_name)
         self.stdout.write("Creating Devices...")
         DeviceFactory.create_batch(20, using=db_name)
+        self.stdout.write("Creating SoftwareImages without Devices or DeviceTypes...")
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/core/management/commands/generate_test_data.py:170]: without. The Device and DeviceType factories automatically assign SoftwareImages and SoftwareVersions (this is a typo I'll fix)

File: nautobot/core/tests/test_forms.py [hunk 21]
```diff
@@ -514,11 +514,11 @@ def setUp(self):
...
         self.namespace = ipam_models.Namespace.objects.first()
-        self.prefix = ipam_models.Prefix.objects.create(
-            prefix="10.0.0.0/24", namespace=self.namespace, status=prefix_status
+        self.prefix, _ = ipam_models.Prefix.objects.get_or_create(
+            prefix="10.0.0.0/24", namespace=self.namespace, defaults={"status": prefix_status}
         )
-        self.ip = ipam_models.IPAddress.objects.create(
-            address="10.0.0.1/32", namespace=self.namespace, status=ipaddr_status
+        self.ip, _ = ipam_models.IPAddress.objects.get_or_create(
+            address="10.0.0.1/32", parent=self.prefix, defaults={"status": ipaddr_status}
```
[the remark this answers is not in the document]
Comment (gsnider2195) [reply nautobot/core/tests/test_forms.py:521]: Because the factory was randomly creating a "10.0.0.1/32" in this namespace so this was failing with a uniqueness error.

File: nautobot/dcim/models/devices.py (hunk 10 above)
Comment (gsnider2195) [reply nautobot/dcim/models/devices.py:1085]: I brought this over from DLM. My guess is that since a sha512 checksum is 128 characters this was to future proof?

File: nautobot/dcim/tests/test_models.py (hunk 14 above)
Comment (gsnider2195) [reply nautobot/dcim/tests/test_models.py:1202]: It's on the DeviceTypeToSoftwareImage model [here](https://github.com/nautobot/nautobot/pull/5137/files#diff-2c74010b539b5fc10880ede3276b50f7af7309db6161e970293b265eebc83a47R116-R123)

File: nautobot/dcim/templates/dcim/softwareimage_retrieve.html (hunk 4 above)
Comment (glennmatthews) [reply nautobot/dcim/templates/dcim/softwareimage_retrieve.html:7]: Brainstorming, what if we put everything in `left_page` *except* the checksum, which goes at the bottom of the page in `full_width_page`? That's the main one that's problematic as a half-column, right?

File: nautobot/dcim/tests/test_models.py (hunk 14 above)
Comment (gsnider2195) [reply nautobot/dcim/tests/test_models.py:1202]: There's no way to set that property through the ORM from the `DeviceType.software_images` or `SoftwareImage.device_types` today. You have to go directly to the `DeviceTypeToSoftwareImage` model where the validation is already implemented.

File: nautobot/dcim/models/devices.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/dcim/models/devices.py:1056]: That's fair. I thought maybe we were using it in a view or table.

**Review by glennmatthews — APPROVED** [review]
:shipit:

**Review by HanlinMiao — APPROVED** [review]

File: nautobot/dcim/api/urls.py [hunk 22]
```diff
@@ -70,6 +70,11 @@
 # Device Redundancy Group
 router.register("device-redundancy-groups", views.DeviceRedundancyGroupViewSet)
 
+# Software images
+router.register("software-images", views.SoftwareImageViewSet)
+router.register("software-versions", views.SoftwareVersionViewSet)
+router.register("device-type-to-software-image", views.DeviceTypeToSoftwareImageViewSet)
```
Comment (glennmatthews) [inline nautobot/dcim/api/urls.py:76]: Given our convention of pluralizing the URL path strings, should this be something like `device-types-to-software-images` instead?

Comment (gsnider2195) [reply nautobot/dcim/api/urls.py:76]: In IPAM we have `/api/ipam/ip-address-to-interface/` but then we have `/api/dcim/interface-redundancy-group-associations/`. I think either the double plural or the `associations` makes it more apparent that you're dealing with an m2m. What do you think?

Comment (glennmatthews) [reply nautobot/dcim/api/urls.py:76]: I think `ip-address-to-interface` also feels "wrong" to me. I'd vote for the double plural.

Comment (gsnider2195) [reply nautobot/dcim/api/urls.py:76]: And I forgot to fix this before merging 😦 
I opened #5190 to fix this
